src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs

```diff
@@ -0,0 +1,619 @@
+            bool found = false;
+            var props = classDeclaration.DescendantNodes().OfType<PropertyDeclarationSyntax>().Where(PropertyIsPublic);
+            foreach (var prop in props)
+            {
+                var loc = prop.GetLocation();
+                var propId = prop.Identifier;
+                var classId = classDeclaration.Identifier;
+
+                var qualName = prop.DescendantNodes().OfType<QualifiedNameSyntax>(); 
+                found |= SignatureContainsTypeName(ref context, qualName, "System.Array", classId, loc, propId);
+
+                var idName = prop.DescendantNodes().OfType<IdentifierNameSyntax>(); 
+                found |= SignatureContainsTypeName(ref context, idName, "Array", classId, loc, propId);
+                
+                found |= ArrayIsntOneDim(prop.DescendantNodes().OfType<ArrayTypeSyntax>(), ref context, classId, propId, loc);
+            }
+            return found;
+        }
+
+        /// <summary>
+        /// Look at all the array types and if any are of the form [][]+ or [,+] then raise the corresponding diagnostic and return true
+        /// </summary>
+        /// <param name="arrTypes"></param>
+        /// <param name="context"></param>
+        /// <param name="typeIdentifier">The type the array lives in</param>
+        /// <param name="fieldId">The code the array is a part of the signature for; e.g. property or method</param>
+        /// <param name="loc"></param>
+        /// <returns>True iff any of the array types given are multidimensional or jagged</returns>
+        private bool ArrayIsntOneDim(IEnumerable<ArrayTypeSyntax> arrTypes, ref GeneratorExecutionContext context, SyntaxToken typeIdentifier, SyntaxToken fieldId, Location loc)
+        {
+            foreach (var arrType in arrTypes)
+            {
+                var brackets = arrType.DescendantNodes().OfType<ArrayRankSpecifierSyntax>();
+                if (brackets.Count() > 1) // [][]+ ?
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(DiagnosticRules.ArraySignature_JaggedArrayRule, loc, fieldId, typeIdentifier));
+                    return true; // could do or-eq on a `found` boolean instead of exiting as soon as we see invalid...
+                }
+                // [,_] ? 
+                else if (brackets.Count() == 1 && brackets.First().ToString().Contains(","))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(DiagnosticRules.ArraySignature_MultiDimensionalArrayRule, loc, fieldId, typeIdentifier));
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        #endregion
+
+        #region Parameters
+
+        private bool ParamHasInOrOutAttribute(ParameterSyntax param)
+        {
+            string baseString = ""; //"System.Runtime.InteropServices.";
+            return ParamHasAttribute(baseString + "In", param) || ParamHasAttribute(baseString + "Out", param);

```

Comment:
Do we need to handle case where it has namespace and where it doesn't?  maybe we want to do a symbol check rather than name check?

Reply:
I changed it but didn't change the check style, let me know what you think